src/livestatus.c

```diff
@@ -0,0 +1,463 @@
+      return 1;
+    }
+
+    livestatus_obj.max_retry = atoi(value);
+  }
+
+  if (strcasecmp("OnFailureBackOffSeconds", key) == 0) {
+    char *backoff_str = strdup(value);
+    if (backoff_str == NULL) {
+      ERROR("livestatus plugin: strdup failed: %s", STRERRNO);
+      return 1;
+    }
+
+    livestatus_obj.backoff_sec = atoi(value);
+  }
+
+  return 0;
+} /* static int ls_config */
+
+static int unix_connect(const char *sockfile, int *sockfd) {
+  struct sockaddr_un sun;
+  int rc = -1;
+  int sfd = -1;
+
+  sfd = socket(AF_UNIX, SOCK_STREAM, 0);
+  if (sfd < 0) {
+    ERROR("livestatus plugin: creating client socket: %s", STRERRNO);
+    return -1;
+  }
+
+  memset(&sun, 0x0, sizeof(sun));
+  sun.sun_family = AF_UNIX;
+  strcpy(sun.sun_path, sockfile);
+
+again:
+  rc = connect(sfd, (struct sockaddr *)&sun, sizeof(sun));
+  if (rc < 0) {
+    if (errno != EINTR) {
+      ERROR("livestatus plugin: connect to socket file %s: %s", sockfile,
+            STRERRNO);
+      return -1;
+    }
+    goto again;
+  }
+
+  *sockfd = sfd;
+
+  return 0;
+} /* unix_connect */
+
+static int ls_parse(const char *lresponse, livestatus_status_t *lstatus) {
+  int rc = -1;
+  int i = 0;
+  char **fields = NULL;
+  char *pchar = NULL;
+
+  fields = malloc(sizeof(char **) * LIVESTATUS_EXPECTED_FIELDS_RESP_NB);
+  if (fields == NULL) {
+    ERROR("livestatus plugin: malloc failed: %s", STRERRNO);
+    return -1;
+  }
+
+  for (int j = 0; j < LIVESTATUS_EXPECTED_FIELDS_RESP_NB; j++) {
+    fields[j] = malloc(sizeof(char *) * 32);
+    if (fields[j] == NULL) {
+      ERROR("livestatus plugin: malloc failed: %s", STRERRNO);
+      rc = -1;
+      goto free_all_fields;
+    }
+
+    memset(fields[j], 0x0, 32);
+  }
+
+  pchar = (char *)lresponse;
+  char *wstart = (char *)lresponse;
+  int exit = 0;
+  while (!exit) {
+    if (*pchar == '\0' || *pchar == '\n') {
+      exit = 1;
+    } else {
+      if (*pchar != ';') {
+        pchar++;
+        continue;
+      }
+    }
+
+    strncpy(fields[i], wstart, pchar - wstart);
+    wstart = pchar + 1;
+    i++;
+    pchar++;
+
+    if (!exit) {
+      if (i >= LIVESTATUS_EXPECTED_FIELDS_RESP_NB) {
+        ERROR("livestatus plugin: too many fields in livestatus output");
+        rc = -2;
+        goto free_all_fields;
+      }
+    }
+  }
+
+  if (i < LIVESTATUS_EXPECTED_FIELDS_RESP_NB) {
+    ERROR("livestatus plugin: not enough fields in livestatus output");
+    rc = -3;
+    goto free_all_fields;
+  }
+
+  *lstatus = c_to_livestatus_status((const char **)fields);
+
+  rc = 0;
+
+free_all_fields:
+  for (int j = 0; j < LIVESTATUS_EXPECTED_FIELDS_RESP_NB; j++) {
+    if (fields[j] != NULL) {
+      free(fields[j]);
+      fields[j] = NULL;
+    }
+  }
+
+  free(fields);
+  fields = NULL;
+
+  return rc;
+} /* int ls_parse */
+
+static int ls_read_parse(const int sockfd, livestatus_status_t *lstatus) {
+  ssize_t bread = -1;
+  char buffer[IO_BUFFER_SIZE];
+
+again:
```

Comment:
goto is sometimes used for error handling, but here it is more typical use for a while loop.

Reply:
I've migrated to a `do{...}while(errno == EINTR)` pattern. Is this okay for you ?